components/core/src/clp/hash_utils.cpp

```diff
@@ -0,0 +1,217 @@
+#include "hash_utils.hpp"
+
+#include <memory>
+#include <span>
+#include <string>
+#include <vector>
+
+#include <fmt/format.h>
+#include <openssl/err.h>
+#include <openssl/hmac.h>
+#include <openssl/sha.h>
+
+#include "ErrorCode.hpp"
+#include "TraceableException.hpp"
+
+using std::make_unique;
+using std::span;
+using std::string;
+using std::unique_ptr;
+using std::vector;
+
+namespace clp {
+namespace {
+/**
+ * Gets the OpenSSL error string as a c++ string
+ * @return The string representing the OpenSSL error
+ */
+auto get_openssl_error_string() -> string {
+    auto const openssl_err = ERR_get_error();
+    return {ERR_error_string(openssl_err, nullptr)};
```

Comment:
Thanks for catching this. In the nullptr case, should we return some customzied error message so users don't get a "" which could be quite confusing?

Reply:
Uh, we could. The reason I chose `{}` is because if you call this method and no error had occurred, then you would also get `nullptr`. So perhaps we should do, in order:

* `0 == openssl_err` => `{}`
* `nullptr == openssl_err_str` => `{"Unknown error"}`
* else => `{openssl_err_str}`

Reply:
Sorry, on no error, `ERR_error_string` does return a string. So I guess we can just do the last two bullets.

Reply:
Perhaps to be clearer, we can use `"Error has no string representation"` instead of `"Unknown error"`.

Reply:
how about 
`"Error has no string representation, error_code: " + std::to_string(openssl_err)`?

Reply:
Sg